DIPs/DIPNNNN.md

```diff
@@ -0,0 +1,349 @@
+
+1. Whenever compiler has to do a symbol lookup, if has to check if *future
+   symbol* is present in the result. If present, deprecation message has to be
+   printed and everything else should continue as if it wasn't present.
+
+   ```D
+   import moda; // provides foo
+   import modb; // provides @future foo
+
+   foo(); // Deprecation: upcoming adddition of modb.foo will result in a symbol
+          // clash, please disambugate
+   ```
+
+2. Whenevr compiler hasto check for presence of a symbol, if there is a match
+   which is *future symbol*, deprecation message has to be printed and
+   everything else should continue as it wasn't present.
+
+   ```D
+   class A
+   {
+       void foo() @future;
+   }
+
+   class B
+   {
+       void foo(); // Deprecation: upcoming addition of A.foo will result in a
+                   // symbol clash, please adjust
+   }
+   ```
+
+3. References to *future symbol* within the module it is declared in should not
+   result in any deprecation messages because the module can be adjusted in the
+   same time symbol is actually added.
+
+   ```D
+   void foo (long a);
+   void foo (int a) @future;
+
+   foo(42); // No point in printing deprecation here as it will keep compiling
+            // even when @future is removed
+   ```
+
+4. Any form of static reflection that provides sequence of results, for example
+   `__traits(allMembers)` or `.tupleof`, has to ignore any *future symbol*
+   completely. Such code does not normally rely on presence of any specific
+   symbol and thus addition of new one is unlikely to cause problems. On the
+   other hand, issuing deprecation in such context will result in a warning that
+   cannot be addressed by adjusting the code.
+
+5. Any form of static reflection that implies access to specific named symbol,
+   for example, `__traits(getMember)`, should still print the deprecation
+   message.
+
+As first implementation step this functionality should only be available as a
+hard list of symbols built into compiler itself and reserved for druntime
+usage. See [breaking changes section](#breaking-changes--deprecation-process)
+for detailed reasoning.
+
+Once concept proves itself with druntime, it should be proposed as an actual
+language feature for mass usage - by either updating this DIP or proposing new
+one. For example, one can simply expose it with new attribute, `@future`.
+
+### Rationale
+
+1. Currently there is no way to warn users of a library about planned addition
+   of new symbol.
+2. Existing workarounds don't eliminate risk of breakage completely and all
+   suffer from some practical drawbacks.
+3. Described problems make it very hard to add new symbols to
+   `object.d` because it is implicitly imported in all D module of all D
+   projects, increasing the chance of name clash as D userbase grows.
+4. Proposed change is not argued to be of notable importance for many D
+   projects. Instead, it is proposed on the basis of being very important for
```

Comment:
s/Proposed/The proposed/
s/is not argued to be of notable importance/is not deemed of high importance/